src/stan_math_backend/Transform_Mir.ml

```diff
@@ -240,24 +305,72 @@ let trans_prog (p : typed_prog) =
         true
     | _ -> false
   in
+  let translate_to_open_cl stmts =
+    if use_opencl then
+      let data_var_idents = List.map ~f:fst p.input_vars in
+      let rec trans_stmt_to_opencl s =
+        { s with
+          stmt=
+            map_statement
+              (switch_expr_to_opencl data_var_idents)
```

Comment:
This is where we say whether the var is available or not - it's only the data variables that we're sending.

I was actually thinking this might be okay for now. Some considerations:
1) This is only a bad encoding if this is done multiple times in a single `log_prob` call, right? In the simple GLM case it isn't really.
2) Once we hook up the optimizer, I would hope to see these "common subexpressions" eliminated and pulled up to a point in log_prob such that `to_matrix_cl(X)` is only called once. I think the optimizer is hooked up now so I'd be curious if it's doing that...
3) Eventually, hopefully our optimizer is smart enough to pull `to_matrix_cl` calls on data out from the log_prob into the prepare_data / constructor block, but for now I don't think that's implemented so I think what we're doing for data now makes sense

Reply:
For this model we should not be using matrix_cls
```
data {
  int<lower=1> k;
  int<lower=0> n;  
  vector[n] y;
}

parameters {
  vector[k] beta;
  real<lower=0> sigma;
  real alpha;
  matrix[n, k] X;
}

model {
  target += normal_id_glm_lpdf(y | X, alpha, beta, sigma);
}
```
But we still do, the above example `lp_accum__.add(normal_id_glm_lpdf<false>(y_opencl__, to_matrix_cl(X), alpha, beta, sigma));` is from this model. 

I think @t4c1 told me that if y was a parameter we would maybe be fine using a GPU, or the other way around, it depends on the GLM we use. But the speedups are not big in those cases.

Reply:
Ohh, I misunderstood the issue. So what are the conditions under which we should use normal_id_glm? Is it just if X (2nd argument) is data?

Reply:
I think X is mandatory to be data for all glms to move to opencl overload.

Reply:
And if 1st arg is already in opencl, we should switch to that version, right?

Reply:
Yes, we dont want to move it just for this function. We would be fine if the first version of this would use gpu glm only if x and y are both data. if that simplifies the problem. And we then go from there.

Reply:
The decision rule for these 4 GLMs is simple: use opencl if and only if X is already on GPU.

For GLMs added [here](https://github.com/stan-dev/stanc3/pull/313#discussion_r330023209) we need different rules. `categorical_logit_glm_lpmf` should use opencl if X is a matrix (and not a row vector), regerdless of whether it is already on GPU or not. `ordered_logistic_glm_lpmf` should use opencl if X is a matrix and it is already on GPU. This rule will also be used for first four GLMs once broadcasting support for them is implemented.

Reply:
Based on Tadej's comment lets go with: "use matrix_cl overload if at least X (2nd arg) is data" (which also means its already on the GPU). I will try to see if I can make this work otherwise I will call for help.

Reply:
I am not sure I follow. 

These 2 could use GPU GLM in all cases (x data, y data, x & y data, no data)?

```
real categorical_logit_glm_lpmf(int, matrix, vector, matrix)
real categorical_logit_glm_lpmf(int[], matrix, vector, matrix)
```
What about these 2?
```
real categorical_logit_glm_lpmf(int, row_vector, vector, matrix)
real categorical_logit_glm_lpmf(int[], row_vector, vector, matrix)
```

Reply:
Actually we could also use opencl in this case if beta (last arguemnt) is already on GPU.